src/coreclr/jit/morph.cpp

```diff
@@ -10115,8 +10115,14 @@ GenTree* Compiler::fgOptimizeCastOnStore(GenTree* store)
     if (!src->OperIs(GT_CAST))
         return store;
 
-    if (store->OperIs(GT_STORE_LCL_VAR) && !lvaGetDesc(store->AsLclVarCommon())->lvNormalizeOnLoad())
-        return store;
+    if (store->OperIs(GT_STORE_LCL_VAR))
+    {
+        LclVarDsc* varDsc = lvaGetDesc(store->AsLclVarCommon()->GetLclNum());
+
+        // It is not safe to remove the cast for non-NormalizeOnLoad variables, parameters or struct fields.
+        if (!varDsc->lvNormalizeOnLoad() || varDsc->lvIsParam || varDsc->lvIsStructField)
```

Comment:
This covers all normalize-on-load cases except OSR locals and address exposure cases. It should likely cover OSR locals as well. Additionally we should have a better comment, like @SingleAccretion has mentioned. Something like:

```suggestion
        // We can make this transformation only under the assumption that NOL locals are always normalized before they are used,
        // however this is not always the case: the JIT will utilize subrange assertions for NOL locals to make normalization
        // assumptions -- see fgMorphLeafLocal. Thus we can only do this for cases where we know for sure that subsequent uses
        // will normalize, which we can only guarantee when the local is address exposed.
        if (!varDsc->lvNormalizeOnLoad() || !varDsc->IsAddressExposed())
```

With this, I'd like to know what the benefit of this transformation even is. The address exposure case is likely very rare compared to the other cases that we are now disabling. How much do we actually gain from having this transformation with these restrictions?

Reply:
(I still question the logic in `fgMorphLeafLocal` -- I feel like it is nonsensical to rely on subrange assertions in the case that it does, which tightly couples it to implementation details of the backend. But that's probably something we can revisit in .NET 9, if we don't manage to find a better fix here.)